### PR TITLE
feat(domain): package read queries with SQL-aggregated stats

### PR DESCRIPTION
**Pull request by mpiton**
## Summary

Implements CQRS read-side queries for Package aggregates with SQL-level aggregation (COUNT, SUM, AVG) of member downloads. Unblocks the Package UI (task 29). Fully tested: 16 new integration tests, all acceptance criteria met.

## Why

Packages need pre-computed statistics (download count, total size, progress %) without N+1 query overhead. CQRS separation: PackageReadRepository returns flattened DTOs from a single GROUP BY query, never exposes mutation. Matches established pattern from Accounts queries.

## Changes

- Add `PackageView` and `PackageFilter` domain models (`domain/model/views.rs`)
- Add `PackageReadRepository` port trait (`domain/ports/driven/package_read_repository.rs`) with three methods:
  - `find_packages(filter)` — lists with optional source_type and name_q (fuzzy) filters
  - `find_package_by_id(id)` — returns single package or None
  - `find_package_downloads(id)` — returns member downloads ordered by queue_position
- Implement `SqlitePackageReadRepo` adapter (`adapters/driven/sqlite/package_read_repo.rs`):
  - `PACKAGE_AGG_SELECT` constant: single `LEFT JOIN ... GROUP BY` query with stats aggregation
  - `row_to_view()` for SQL→domain mapping (16 integration tests verify correctness)
  - Handles NULL aggregates correctly (COALESCE for SUM/COUNT edge cases)
- Add three Query handlers: `ListPackagesQuery`, `GetPackageQuery`, `ListPackageDownloadsQuery`
- Add `PackageViewDto` read model with `#[serde(rename_all = "camelCase")]` for frontend
- Wire in QueryBus and expose three Tauri IPC commands: `package_list`, `package_get`, `package_list_downloads`
- Update CHANGELOG.md

## Testing

All acceptance criteria verified:
- **N+1 prevention**: aggregation at SQL level via `LEFT JOIN ... GROUP BY p.id`, not N+1 application queries. Tested by `test_find_packages_aggregates_member_downloads`.
- **Fuzzy name filter**: `LOWER(p.name) LIKE ?` is case-insensitive substring match. Verified by `test_find_packages_filter_by_name_q_is_case_insensitive_substring`.
- **SQLite integration**: 16 tests via `setup_test_db()` + fixtures, no mocks. All pass locally and in CI.

```bash
cargo test package_read_repo -- --nocapture
cargo test --workspace
```

Full test output: 1299 cargo tests pass, 605 vitest pass, clippy clean, cargo fmt clean, oxlint clean.

## Related Issues

- Closes task 28 (Vortex sprint roadmap)
- Unblocks task 29 (Vue Packages React, all dependencies now satisfied)

## Checklist

- [x] Tests added/updated and passing locally
- [x] Docs updated (CHANGELOG.md)
- [x] No secrets, debug prints, or commented-out code
- [x] Self-reviewed the diff
- [x] CI green

---
## Summary by cubic
Add read-side package queries with SQL-aggregated stats to avoid N+1 and unblock the Packages UI (task 28). Also clamps progress, scales large member lookups, suppresses ETAs on Completed, and fixes legacy created_at fallback in package member views.

- **New Features**
  - New `PackageReadRepository` and `SqlitePackageReadRepo` compute `downloads_count`, `total_bytes`, `downloaded_bytes`, `progress_percent`, and `all_completed` via one `LEFT JOIN ... GROUP BY`.
  - Query handlers `ListPackagesQuery`, `GetPackageQuery`, `ListPackageDownloadsQuery` wired into `QueryBus`; DTO `PackageViewDto` (camelCase, no password). IPC: `package_list`, `package_get`, `package_list_downloads`; list results ordered by `(created_at, id)` and member downloads by `queue_position`.
  - Filtering: `source_type` exact match; `name_q` is a Unicode-aware, case-insensitive substring in Rust; both AND together; blank `name_q` ignored.

- **Bug Fixes**
  - Aggregate credits `Completed` members at full `total_bytes` (fallback to `downloaded_bytes` when total is NULL) symmetrically in numerator and denominator.
  - Clamp `progress_percent` to 100% when `downloaded_bytes > total_bytes` for non-completed rows (aggregate and per-download views).
  - Chunk member ID and segment lookups in `find_package_downloads` to stay under SQLite’s variable limit for large packages.
  - Suppress `eta_seconds` for `Completed` downloads.
  - Resolve legacy member `created_at` with a fallback chain (id-inferred → `updated_at` → sentinel) via `resolve_download_created_at`, used in both `package_read_repo` and `download_read_repo` to avoid “1970” dates and keep ordering consistent.

<sup>Written for commit f30fd83cdba242fc0d4e4203961fb7d6f13a419c. Summary will update on new commits. <a href="https://cubic.dev/pr/mpiton/vortex/pull/132?utm_source=github">Review in cubic</a></sup>

## Summary by CodeRabbit

* **New Features**
  * Package browsing: list packages (filter by source type and name substring), deterministic ordering, package details with aggregated download stats/progress, and list package downloads via new IPC endpoints.
  * Accounts: quota-driven account rotation with selector exclusions and related domain/events; account UI and persistence added.
* **Tests**
  * Extensive tests for package aggregation, progress clamping, filtering, ordering, DTO mapping, and query handlers.
* **Documentation**
  * CHANGELOG updated to document these additions.

### CODE REVIEW

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 1]
```diff
@@ -0,0 +1,685 @@
...
+        progress_percent,
+        all_completed,
+    })
+}
+
+const PACKAGE_AGG_SELECT: &str = "SELECT \
+    p.id, p.name, p.source_type, p.folder_path, p.auto_extract, p.priority, p.created_at, \
+    COUNT(d.id) AS downloads_count, \
+    COALESCE(SUM(COALESCE(d.total_bytes, 0)), 0) AS total_bytes_sum, \
+    COALESCE(SUM(d.downloaded_bytes), 0) AS downloaded_bytes_sum, \
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @chatgpt-codex-connector

> The aggregate query computes `downloaded_bytes_sum` as `SUM(d.downloaded_bytes)`, but completed downloads are known to sometimes persist with `downloaded_bytes < total_bytes` (the per-download view explicitly treats `Completed` as 100%).

Fixed in `c40208f`. Switched the SUM to a CASE that credits Completed rows at `COALESCE(total_bytes, downloaded_bytes)`, so the aggregate matches `compute_progress_percent_for_download`'s 100% rule. Two regression tests cover it: the drift case (300/1000 Completed + 250/1000 Downloading now reads 62.5% instead of 27.5%) and the NULL-total Completed fallback that credits `downloaded_bytes`.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+    let progress_percent = compute_progress_percent_for_download(&model.state, downloaded, total);
+    let eta_seconds = match total {
+        Some(t) if speed > 0 && t > downloaded => Some((t - downloaded) / speed),
+        _ => None,
+    };
+    let state = model.state.parse().map_err(|_| {
+        DomainError::StorageError(format!("invalid download state in DB: {}", model.state))
+    })?;
+    let priority_u8 = u8::try_from(model.priority).unwrap_or(5);
+    let created_at = safe_u64(model.created_at);
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @coderabbitai

> Avoid silently coercing invalid `downloads.priority` to `5`. Falling back here hides bad persisted data and can mislead queue/priority UI behavior.

Out of scope for this PR. The `unwrap_or(5)` is a deliberate mirror of `download_read_repo.rs:81` — both reader paths share the same fallback so a malformed priority column behaves identically across the package view and the download view. Tightening it here in isolation would diverge them: a row with `priority = 999` would surface an error in `find_package_downloads` but still render in `find_downloads`. The right fix is to switch both readers to `Result` propagation (and add a storage-side validation) in one pass; tracking it as a follow-up rather than splitting the change here.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 3]
```diff
@@ -0,0 +1,777 @@
...
+            if let Some(ref source) = f.source_type {
+                clauses.push("p.source_type = ?");
+                params.push(Value::from(source.clone()));
+            }
+            if let Some(ref needle) = f.name_q {
+                let trimmed = needle.trim();
+                if !trimmed.is_empty() {
+                    clauses.push("LOWER(p.name) LIKE ?");
+                    lowered_name = Some(format!("%{}%", trimmed.to_lowercase()));
+                    params.push(Value::from(lowered_name.unwrap()));
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @coderabbitai resolve

> Escape `LIKE` metacharacters in `name_q`. This turns the advertised substring filter into SQL pattern matching for inputs containing `%` or `_`.

Fixed in `242fe4a`. Added an `escape_like` helper that escapes `\`, `%`, `_` in user input, and the clause now reads `LOWER(p.name) LIKE ? ESCAPE '\'`. Three regression tests cover the metachars: `100%` no longer matches `100 packages`, `foo_bar` no longer matches `foo-bar` / `fooXbar`, and a trailing `\` no longer corrupts the wrapping `%...%`.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 4]
```diff
@@ -0,0 +1,777 @@
...
+// as 100% regardless of the persisted `downloaded_bytes`. The CASE
+// branch below mirrors that — completed members contribute their full
+// `total_bytes` (or fall back to `downloaded_bytes` when total is
+// unknown) so a package with mixed Completed + active members never
+// reports a lower aggregate than the sum of its per-row progress.
+const PACKAGE_AGG_SELECT: &str = "SELECT \
+    p.id, p.name, p.source_type, p.folder_path, p.auto_extract, p.priority, p.created_at, \
+    COUNT(d.id) AS downloads_count, \
+    COALESCE(SUM(COALESCE(d.total_bytes, 0)), 0) AS total_bytes_sum, \
+    COALESCE(SUM(CASE WHEN d.state = 'Completed' THEN COALESCE(d.total_bytes, d.downloaded_bytes) ELSE d.downloaded_bytes END), 0) AS downloaded_bytes_sum, \
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs:133]: @chatgpt-codex-connector

> This aggregate can produce `progress_percent > 100` for mixed-state packages: when a `Completed` member has `total_bytes = NULL`, `downloaded_bytes_sum` adds its `downloaded_bytes` (line 123) but `total_bytes_sum` still adds `0` (line 122).

Caught a real regression I introduced in `c40208f`. Fixed in `242fe4a` — both sums now share the same per-row CASE for Completed rows (`COALESCE(total_bytes, downloaded_bytes)`), so the NULL-total fallback hits the numerator and the denominator together. Added `test_find_packages_mixed_state_with_unknown_completed_total_stays_under_100` (500 NULL-total Completed + 50/100 Downloading → 91.7%, not 500%) and updated the existing single-member test to assert symmetric credit.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 5]
```diff
@@ -0,0 +1,940 @@
...
+// Keeping the CASE symmetric also mirrors the per-download semantics
+// in `compute_progress_percent_for_download`, which treats Completed
+// as 100% regardless of the persisted bytes.
+const PACKAGE_AGG_SELECT: &str = "SELECT \
+    p.id, p.name, p.source_type, p.folder_path, p.auto_extract, p.priority, p.created_at, \
+    COUNT(d.id) AS downloads_count, \
+    COALESCE(SUM(CASE WHEN d.state = 'Completed' THEN COALESCE(d.total_bytes, d.downloaded_bytes) ELSE COALESCE(d.total_bytes, 0) END), 0) AS total_bytes_sum, \
+    COALESCE(SUM(CASE WHEN d.state = 'Completed' THEN COALESCE(d.total_bytes, d.downloaded_bytes) ELSE d.downloaded_bytes END), 0) AS downloaded_bytes_sum, \
+    COALESCE(SUM(CASE WHEN d.state = 'Completed' THEN 1 ELSE 0 END), 0) AS completed_count \
+    FROM packages p LEFT JOIN downloads d ON d.package_id = p.id";
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs:135]: @coderabbitai resolve

> Update the exported `PackageView.total_bytes` contract. These aggregates now count `downloaded_bytes` for `Completed` rows whose `total_bytes` is `NULL`, but `src-tauri/src/domain/model/views.rs` still says unknown sizes contribute `0`.

Fixed in `1a36e14`. Rewrote both `PackageView.total_bytes` and `PackageView.downloaded_bytes` doc comments in `domain/model/views.rs` to spell out the symmetric Completed-row CASE: members in state `Completed` count for `COALESCE(total_bytes, downloaded_bytes)`, other members keep the old semantics. The `progress_percent` invariant (`<= 100%`) is documented as the reason for the symmetry.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 6]
```diff
@@ -0,0 +1,940 @@
...
+                let trimmed = needle.trim();
+                if !trimmed.is_empty() {
+                    // ESCAPE clause keeps the advertised "substring"
+                    // semantics: a literal `%` or `_` in user input
+                    // matches itself instead of acting as a wildcard.
+                    clauses.push(r"LOWER(p.name) LIKE ? ESCAPE '\'");
+                    params.push(Value::from(format!(
+                        "%{}%",
+                        escape_like(&trimmed.to_lowercase()),
+                    )));
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @coderabbitai resolve

> `LOWER(...) LIKE` fails for non-ASCII Unicode case-insensitive matches in stock SQLite. `LOWER('CAFÉ')` returns `'cafÉ'`, not `'café'`.

Fixed in `1a36e14`. Dropped the SQL `LOWER(p.name) LIKE ? ESCAPE '\'` clause and the `escape_like` helper entirely. The `name_q` filter now runs in Rust after the SQL fetch using `name.to_lowercase().contains(&needle.to_lowercase())` — Rust's `to_lowercase` is Unicode-aware, so `CAFÉ` matches `café` and `Ärger` matches `ärger`. Same approach `history_repo` already uses (see comment at `history_repo.rs:147`). New regression test `test_find_packages_filter_name_q_matches_non_ascii_case_insensitively` covers the cases from your sqlite repro. The metachar-literal tests still pass (and were renamed to drop `_escapes_` since literalness now comes from `str::contains`, not a SQL ESCAPE clause).

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 7]
```diff
@@ -0,0 +1,994 @@
...
+
+fn aggregate_progress_percent(downloaded: u64, total: u64, all_completed: bool) -> f64 {
+    if all_completed {
+        return 100.0;
+    }
+    if total == 0 {
+        return 0.0;
+    }
+    round_one_dp(downloaded as f64 / total as f64 * 100.0)
+}
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs:52]: @coderabbitai resolve

> Clamp computed progress to 100.0 to match the documented contract. On Line 45 and Line 136, ratios can exceed `100.0` when persisted `downloaded_bytes > total_bytes` for non-completed rows.

Fixed in `eceacd2`. Added `.min(100.0)` to both `aggregate_progress_percent` and `compute_progress_percent_for_download`. Regression test `test_find_packages_progress_clamps_above_one_hundred_percent` seeds a Downloading row at 1500/1000 and asserts both the package aggregate and the per-row view return exactly 100.0.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 8]
```diff
@@ -0,0 +1,994 @@
...
+                return Ok(Vec::new());
+            }
+
+            let download_ids: Vec<i64> = id_rows
+                .iter()
+                .map(|r| r.try_get_by_index::<i64>(0).map_err(map_db_err))
+                .collect::<Result<Vec<_>, _>>()?;
+
+            let downloads = download::Entity::find()
+                .filter(download::Column::Id.is_in(download_ids.clone()))
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @chatgpt-codex-connector

> When a package has a large number of member downloads, `is_in(download_ids.clone())` expands to one SQL bind parameter per id; this can exceed SQLite's host-parameter limit (build-dependent, commonly 999 or 32766) and make `find_package_downloads` fail with a storage error.

Fixed in `eceacd2`. Added `SQLITE_IN_CHUNK = 900` (well below the 999 floor) and chunk both the `download::Entity::find().filter(Id.is_in(...))` lookup and the `download_segment` lookup with `download_ids.chunks(SQLITE_IN_CHUNK)`. The order-preserving `by_id` HashMap merge step was already chunk-agnostic. Regression test `test_find_package_downloads_chunks_oversized_in_query` seeds `SQLITE_IN_CHUNK + 50` rows in one package and asserts every member comes back in queue order.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 9]
```diff
@@ -0,0 +1,1109 @@
...
+    segments_total: u32,
+) -> Result<DownloadView, DomainError> {
+    let total = model.total_bytes.map(safe_u64);
+    let downloaded = safe_u64(model.downloaded_bytes);
+    let speed = safe_u64(model.speed_bytes_per_sec);
+    let progress_percent = compute_progress_percent_for_download(&model.state, downloaded, total);
+    let eta_seconds = match total {
+        Some(t) if speed > 0 && t > downloaded => Some((t - downloaded) / speed),
+        _ => None,
+    };
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @coderabbitai resolve

> Suppress ETA for completed downloads. A `Completed` row can still get a non-`None` `eta_seconds` here if stale persisted data leaves `speed_bytes_per_sec > 0` and `downloaded_bytes < total_bytes`.

Fixed in `cbf5fa4`. Wrapped the existing `total / speed / downloaded` match in an explicit `if model.state == "Completed" { None }` guard, so the ETA suppression keys off state instead of trusting the byte counters. Regression test `test_find_package_downloads_eta_is_none_for_completed_with_stale_speed` seeds a Completed row at 600/1000 with `speed = 100` (would naively compute 4s) and asserts `eta_seconds == None`, while a Downloading sibling with the same numbers still gets `Some(4)`.

File: src-tauri/src/adapters/driven/sqlite/package_read_repo.rs [hunk 10]
```diff
@@ -0,0 +1,1167 @@
...
+        match total {
+            Some(t) if speed > 0 && t > downloaded => Some((t - downloaded) / speed),
+            _ => None,
+        }
+    };
+    let state = model.state.parse().map_err(|_| {
+        DomainError::StorageError(format!("invalid download state in DB: {}", model.state))
+    })?;
+    let priority_u8 = u8::try_from(model.priority).unwrap_or(5);
+    let created_at = safe_u64(model.created_at);
```
[the remark this answers is not in the document]
Comment (mpiton) [reply src-tauri/src/adapters/driven/sqlite/package_read_repo.rs]: @chatgpt-codex-connector

> `download_row_to_view` sets `created_at` directly from `model.created_at`, but legacy rows can persist `created_at = 0` and `SqliteDownloadReadRepo` explicitly normalizes those via id/updated_at fallback.

Fixed in `f30fd83`. Extracted the fallback chain from `download_read_repo::read_created_at` into a shared `resolve_download_created_at(raw_ca, raw_id, raw_ua)` helper in `util.rs` (takes primitives so `util` does not depend on entity types) and routed both readers through it. Regression test `test_find_package_downloads_resolves_legacy_zero_created_at` covers the three fallback branches: snowflake-id inference (1_700_000_000_000 << 12 → recovered), `updated_at` fallback when the id has no embedded ms, and the `MIN_PLAUSIBLE_UNIX_MS` sentinel when nothing is salvageable.
